[GTK] Use the buffer format preferred by the driver in DMA-BUF renderer

carlosgcampos opened the pull request:
#### d3a9830bb67830d77bfbb785df77020fc104aef6
<pre>
[GTK] Use the buffer format preferred by the driver in DMA-BUF renderer
<a href="https://bugs.webkit.org/show_bug.cgi?id=263309">https://bugs.webkit.org/show_bug.cgi?id=263309</a>

Reviewed by Alejandro G. Castro.

We currently use ARGB8888 unconditionally, assuming it will be supported
by all drivers. We should query the driver for the list of supported
formats and modifiers and create the GBM buffer with modifiers to make
sure the optimal format and modifier is used. Now it&apos;s possible that the
selected format and modifier results in a multiplane buffer, so this
patch also implements the support for importing and exporting simple
multiplane DMA-BUF buffers.

* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::gbmDevice):
(WebCore::PlatformDisplay::dmabufFormats):
* Source/WebCore/platform/graphics/PlatformDisplay.h:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/Shared/glib/DMABufRendererBufferFormat.h: Added.
* Source/WebKit/Shared/glib/DMABufRendererBufferFormat.serialization.in: Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::preferredBufferFormats):
(WebKit::AcceleratedBackingStoreDMABuf::BufferEGLImage::create):
(WebKit::AcceleratedBackingStoreDMABuf::BufferEGLImage::BufferEGLImage):
(WebKit::AcceleratedBackingStoreDMABuf::BufferGBM::create):
(WebKit::AcceleratedBackingStoreDMABuf::BufferGBM::BufferGBM):
(WebKit::AcceleratedBackingStoreDMABuf::BufferGBM::surface const):
(WebKit::AcceleratedBackingStoreDMABuf::didCreateBuffer):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.h:
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::preferredBufferFormats const):
* Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::AcceleratedSurfaceDMABuf):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetEGLImage::create):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetEGLImage::RenderTargetEGLImage):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetTexture::create):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetTexture::RenderTargetTexture):
(WebKit::AcceleratedSurfaceDMABuf::SwapChain::setupBufferFormat):
(WebKit::AcceleratedSurfaceDMABuf::SwapChain::createTarget const):
* Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.h:
* Source/cmake/OptionsGTK.cmake:

Canonical link: <a href="https://commits.webkit.org/269634@main">https://commits.webkit.org/269634@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/10bc9bfa7494c8927f0dc8f870ad97c75580c35b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23069 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1151 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24183 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24984 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21335 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23335 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2429 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23604 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22193 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23310 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/624 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20006 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25835 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/491 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20891 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27061 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/20069 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20885 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21156 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24929 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/22435 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/574 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18358 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/29807 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/501 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6161 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/982 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/29759 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2938 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/775 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6055 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

alexgcastro commented on Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -116,63 +120,107 @@ void AcceleratedSurfaceDMABuf::RenderTargetColorBuffer::willRenderFrame() const
-        };
-        attributes.append(std::span<const EGLAttrib> { modifierAttributes });
+
+#define ADD_PLANE_ATTRIBUTES(planeIndex) { \
```
Is this define duplicated? Could we use some shared definition?

carlosgcampos replied:
No, it's the same name, but it has a different implementation. One is used in the code to export the DMA-BUF and the other in the code to import it. In both cases the macro is only used locally in that function and unset right after being used.

alexgcastro commented on Source/WebCore/platform/graphics/PlatformDisplay.cpp:
```diff
@@ -527,8 +532,61 @@ struct gbm_device* PlatformDisplay::gbmDevice()
+        if (!s_eglQueryDmaBufFormatsEXT(m_eglDisplay, formatsCount, reinterpret_cast<EGLint*>(formats.data()), &formatsCount))
+            return;
+
+        static PFNEGLQUERYDMABUFMODIFIERSEXTPROC s_eglQueryDmaBufModifiersEXT = extensions.EXT_image_dma_buf_import_modifiers ?
+            reinterpret_cast<PFNEGLQUERYDMABUFMODIFIERSEXTPROC>(eglGetProcAddress("eglQueryDmaBufModifiersEXT")) : nullptr;
```
For s_eglQueryDmaBufFormatsEXT we are using a different way of checking if the extension is present and the early return, shouldn't we use the same check here and early return instead of doing it later?

carlosgcampos replied:
The format is required, but the modifiers aren't. We don't early return if `s_eglQueryDmaBufFormatsEXT` is not present, but only in case of failure calling it. When modifiers are not supported we use a modifiers vector with a single element that is `DRM_FORMAT_MOD_INVALID`.

alexgcastro commented on Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.h:
```diff
@@ -132,7 +133,7 @@ class AcceleratedSurfaceDMABuf final : public AcceleratedSurface, public IPC::Me
     public:
         static std::unique_ptr<RenderTarget> create(uint64_t, const WebCore::IntSize&);
-        RenderTargetTexture(uint64_t, const WebCore::IntSize&, WTF::UnixFileDescriptor&&, unsigned, uint32_t format, uint32_t offset, uint32_t stride, uint64_t modifier);
+        RenderTargetTexture(uint64_t, const WebCore::IntSize&, unsigned, uint32_t format, Vector<WTF::UnixFileDescriptor>&&, Vector<uint32_t>&& offsets, Vector<uint32_t>&& strides, uint64_t modifier);
```
Adding the name texture to the unsigned parameter I think it is a good addition for understanding.

alexgcastro reviewed (approved):
Great improvement in the formats handling! Thanks!

carlosgcampos committed:
[GTK] Use the buffer format preferred by the driver in DMA-BUF renderer
https://bugs.webkit.org/show_bug.cgi?id=263309

Reviewed by Alejandro G. Castro.

We currently use ARGB8888 unconditionally, assuming it will be supported
by all drivers. We should query the driver for the list of supported
formats and modifiers and create the GBM buffer with modifiers to make
sure the optimal format and modifier is used. Now it's possible that the
selected format and modifier results in a multiplane buffer, so this
patch also implements the support for importing and exporting simple
multiplane DMA-BUF buffers.

* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::gbmDevice):
(WebCore::PlatformDisplay::dmabufFormats):
* Source/WebCore/platform/graphics/PlatformDisplay.h:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/Shared/glib/DMABufRendererBufferFormat.h: Added.
* Source/WebKit/Shared/glib/DMABufRendererBufferFormat.serialization.in: Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::preferredBufferFormats):
(WebKit::AcceleratedBackingStoreDMABuf::BufferEGLImage::create):
(WebKit::AcceleratedBackingStoreDMABuf::BufferEGLImage::BufferEGLImage):
(WebKit::AcceleratedBackingStoreDMABuf::BufferGBM::create):
(WebKit::AcceleratedBackingStoreDMABuf::BufferGBM::BufferGBM):
(WebKit::AcceleratedBackingStoreDMABuf::BufferGBM::surface const):
(WebKit::AcceleratedBackingStoreDMABuf::didCreateBuffer):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.h:
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::preferredBufferFormats const):
* Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::AcceleratedSurfaceDMABuf):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetEGLImage::create):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetEGLImage::RenderTargetEGLImage):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetTexture::create):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetTexture::RenderTargetTexture):
(WebKit::AcceleratedSurfaceDMABuf::SwapChain::setupBufferFormat):
(WebKit::AcceleratedSurfaceDMABuf::SwapChain::createTarget const):
* Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.h:
* Source/cmake/OptionsGTK.cmake:

Canonical link: https://commits.webkit.org/269634@main
